core/commonMain/src/ArgParser.kt

```diff
@@ -342,24 +371,116 @@ open class ArgParser(
+        // Boolean flags.
+        if (argValue.descriptor.fullName == "help") {
+            println(makeUsage())
+            exitProcess(0)
+        }
+        saveAsOption(argValue, "true")
+    }
+
+    /**
+     * Save option described with standard separated form `--name value`.
+     *
+     * @param argValue argument value with all information about option.
+     * @param argIterator iterator over command line arguments.
+     */
+    private fun saveStandardOptionForm(argValue: ParsingValue<*, *>, argIterator: Iterator<String>) {
+        if (argValue.descriptor.type.hasParameter) {
+            if (argIterator.hasNext()) {
+                saveAsOption(argValue, argIterator.next())
+            } else {
+                // An error, option with value without value.
+                printError("No value for ${argValue.descriptor.textDescription}")
+            }
+        } else {
+            saveOptionWithoutParameter(argValue)
+        }
+    }
 
     /**
-     * Try to recognize command line element as short form of option.
+     * Try to recognize and save command line element as short form of option.
      *
      * @param candidate string with candidate in options.
+     * @param argIterator iterator over command line arguments.
      */
-    private fun recognizeOptionShortForm(candidate: String) =
-            if (candidate.startsWith(optionShortFromPrefix))
-                shortNames[candidate.substring(optionShortFromPrefix.length)]
-            else null
+    private fun recognizeAndSaveOptionShortForm(candidate: String, argIterator: Iterator<String>): Boolean {
+        if (!candidate.startsWith(optionShortFromPrefix) ||
+            optionFullFormPrefix != optionShortFromPrefix && candidate.startsWith(optionFullFormPrefix)) return false
+        // Try to find exact match.
+        val option = candidate.substring(optionShortFromPrefix.length)
+        val argValue = shortNames[option]
+        if (argValue != null) {
+            saveStandardOptionForm(argValue, argIterator)
+        } else {
+            if (prefixStyle != OptionPrefixStyle.GNU || option.isEmpty())
+                return false
+
+            // Try to find collapsed form.
+            val firstOption = shortNames["${option[0]}"] ?: return false
+            // Form with value after short form without separator.
+            if (firstOption.descriptor.type.hasParameter) {
+                saveAsOption(firstOption, option.substring(1))
+            } else {
+                // Form with several short forms as one string.
+                val otherBooleanOptions = option.substring(1)
+                saveOptionWithoutParameter(firstOption)
+                for (option in otherBooleanOptions) {
+                    if (shortNames["$option"]?.descriptor?.type?.hasParameter != false) {
+                        printError("Unknown option $optionShortFromPrefix$option in option combination $candidate.")
```

Comment:
Better to distinguish situations when the option letter is unknown and when it corresponds to an option with parameter in order to provide more specific error message.